Sources/OpenAPIRuntime/Interface/Body.swift

```diff
@@ -0,0 +1,535 @@
+        )
+    }
+
+    public convenience init<S: Sequence>(
+        dataChunks: S,
+        length: Length,
+        iterationBehavior: IterationBehavior
+    ) where S.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+
+    public convenience init<C: Collection>(
+        dataChunks: C,
+        length: Length
+    ) where C.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: length,
+            iterationBehavior: .multiple
+        )
+    }
+
+    public convenience init<C: Collection>(
+        dataChunks: C
+    ) where C.Element == DataType {
+        self.init(
+            sequence: .init(WrappedSyncSequence(sequence: dataChunks)),
+            length: .known(dataChunks.map(\.count).reduce(0, +)),
+            iterationBehavior: .multiple
+        )
+    }
+
+    public convenience init(
+        stream: AsyncThrowingStream<DataType, any Error>,
+        length: Body.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init(
+        stream: AsyncStream<DataType>,
+        length: Body.Length
+    ) {
+        self.init(
+            sequence: .init(stream),
+            length: length,
+            iterationBehavior: .single
+        )
+    }
+
+    public convenience init<S: AsyncSequence>(
+        sequence: S,
+        length: Body.Length,
+        iterationBehavior: IterationBehavior
+    ) where S.Element == DataType {
+        self.init(
+            sequence: .init(sequence),
+            length: length,
+            iterationBehavior: iterationBehavior
+        )
+    }
+}
+
+// MARK: - Consuming the body
+
+extension Body: AsyncSequence {
+    public typealias Element = DataType
+    public typealias AsyncIterator = Iterator
+    public func makeAsyncIterator() -> AsyncIterator {
+        if iterationBehavior == .single {
+            lock.lock()
+            defer {
+                lock.unlock()
+            }
+            guard !locked_iteratorCreated else {
+                fatalError(
+                    "OpenAPIRuntime.Body attempted to create a second iterator, but the underlying sequence is only safe to be iterated once."
+                )
+            }
+            locked_iteratorCreated = true
+        }
+        return sequence.makeAsyncIterator()
+    }
+}
+
+// MARK: - Transforming the body
+
+extension Body {
+
+    /// Creates a body where each chunk is transformed by the provided closure.
+    /// - Parameter transform: A mapping closure.
+    /// - Throws: If a known length was provided to this body at
+    /// creation time, the transform closure must not change the length of
+    /// each chunk.
+    public func mapChunks(
```

Comment:
Interesting. What is the use-case for this?

Reply:
A common one: logging, a less common one: streaming encryption/unencryption/zipping/unzipping, etc in a middleware. That's why the result value still needs to be a body.

Outside of middlewares, you can just use the normal existing algorithms on AsyncSequence and get `A<B<...<MySequence>>>`, but in middlewares, the type of the sequence we provide to you, and that you provide back to us, has to be `Body`.